gax/src/main/java/com/google/api/gax/rpc/ApiExceptionFactory.java

```diff
@@ -29,15 +29,103 @@
  */
 package com.google.api.gax.rpc;
 
+import java.util.Iterator;
+import java.util.ServiceLoader;
+
 /**
- * A factory class that returns the corresponding type of exception class from the given status
- * code.
+ * A factory class which returns a corresponding {@link ApiException} for a generic cause.
+ *
+ * <p>You can override the default mapping via the {@link ServiceLoader} pattern by declaring an
+ * implementation of {@link ExceptionTransformation}.
  */
 public class ApiExceptionFactory {
-  private ApiExceptionFactory() {}
+
+  private static final ExceptionTransformation DEFAULT_EXCEPTION_TRANSFORMATION =
+      new IdentityExceptionTransformation();
+  private static final ExceptionTransformation transform = loadExceptionTransformation();
 
   public static ApiException createException(
       Throwable cause, StatusCode statusCode, boolean retryable) {
+    ApiException exception = mapStatusToException(cause, statusCode, retryable);
+    exception = transform.transform(exception);
+    return validate(exception, statusCode);
+  }
+
+  public static ApiException createException(
+      String message, Throwable cause, StatusCode statusCode, boolean retryable) {
+    ApiException exception = mapStatusToException(message, cause, statusCode, retryable);
+    exception = transform.transform(exception);
+    return validate(exception, statusCode);
+  }
+
+  /**
+   * Ensures that we will return a valid exception.
+   *
+   * @throws IllegalStateException if the exception is null.
+   */
+  private static ApiException validate(ApiException exception, StatusCode statusCode) {
+    if (exception == null) {
+      throw new IllegalStateException(
+          "Unable to map "
+              + statusCode
+              + " to exception. Possibly a bad ExceptionTransformation service provider?");
+    }
+    return exception;
+  }
+
+  /**
+   * Provides an interface to allow applying an arbitrary transformation to the default exception
+   * mapping.
+   *
+   * <p>This can be useful to map the generic {@link ApiException} implementations to a client
+   * library specific implementation. For instance, to extract server-provided information from the
+   * gRPC trailers and present it to the user.
+   */
+  public interface ExceptionTransformation {
```

Comment:
Please rename to `ExceptionTransformer` such that it names an object instead of a process.

Reply:
Would prefer for this exception to go into a separate file.